File: js/algo/merge-sort.js
function mergeSort(arr) {
  if (arr.length <= 1) return arr;

  let middle = Math.floor(arr.length / 2);
  let left = arr.slice(0, middle);
  let right = arr.slice(middle, arr.length);

  return merge(mergeSort(left), mergeSort(right));
}

function merge(left, right) {
  let results = [];
  let l = 0;
  let r = 0;
  while (l < left.length && r < right.length) {
    if (left[l] < right[r]) {
      results.push(left[l]);
      l++;
    } else {
      results.push(right[r]);
      r++;
    }
  }

  while (l < left.length) {
    results.push(left[l]);
    l++;
  }

  while (r < right.length) {
    results.push(right[r]);
    r++;
  }
  return results;
}

const arr = [5, 1, 4, 2, 8, 6, 9];
console.log(mergeSort(arr));

File: python/algorithms/merge_sort.py
def merge_sort(arr):
    if len(arr) < 2:
        return arr
    
    mid = len(arr) // 2
    left = arr[:mid]
    right = arr[mid:]

    left = merge_sort(left)
    right = merge_sort(right)

    return merge(left, right)

def merge(left, right):
    sorted_list = []
    len_left = len(left)
    len_right = len(right)
    i,j = 0,0

    while i < len_left and j < len_right:
        if left[i] <= right[j]:
            sorted_list.append(left[i])
            i += 1
        else:
            sorted_list.append(right[j])
            j += 1
    
    if i < len_left:
        for l in range(i, len_left):
            sorted_list.append(left[l])
    if j < len_right:
        for r in range(j, len_right):
            sorted_list.append(right[r])
    return sorted_list

if __name__ == "__main__":
    nums = [7, 2, 1, 6, 8, 5, 3, 4]
    nums = merge_sort(nums)
    print(nums)